rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+In order to make that available to be used, [providers.Interface](https://github.com/opentofu/opentofu/blob/26a77c91560d51f951aa760bdcbeecd93f9ef6b0/internal/providers/provider.go#L109) needs to get the necessary methods and implement those in [GRPCProviderPlugin (V5)](https://github.com/opentofu/opentofu/blob/26a77c91560d51f951aa760bdcbeecd93f9ef6b0/internal/plugin/grpc_provider.go#L31) and [GRPCProviderPlugin (V6)](https://github.com/opentofu/opentofu/blob/26a77c91560d51f951aa760bdcbeecd93f9ef6b0/internal/plugin6/grpc_provider.go#L31).
+
+#### Configuration model
+Beside the attributes that are defined by the provider for an ephemeral resource, the following meta-arguments needs to be supported by any ephemeral block:
+* lifecycle
+* count
+* for_each
+* depends_on
+* provider
+
+#### `Open` method details
+When OpenTofu will have to use an ephemeral resource, it needs to call its `Open` method, passing over the config of the ephemeral resource.
+
+The call to the `Open` method will return the following data:
+* `Private` that OpenTofu is not going to use in other contexts than calling the provider `Close` or `Renew` optionally defined methods.
+* `Result` that will contain the actual ephemeral information. This is what OpenTofu needs to handle to make it available to other ephemeral contexts to reference.
+* `RenewAt` being an optional timestamp indicating when OpenTofu will have to call `Renew` method on the provider before using again the data from the `Result`.
+
+Observations:
+* In the `Result`, OpenTofu is epecting to find any non-computed given values in the request, otherwise will return an error.
+* In the `Result`, the fields marked as computed can be either null or have an actual value. If an unknown if found, OpenTofu will return an error.
+
+> [!NOTE]
+>
+> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.
+
+#### `Renew` method details
+The `Renew` method is called only if the response from `Open` or another `Renew` call is containing a `RenewAt`.
+When `RenewAt` is present, OpenTofu, before using the `Result` from the `Open` method response, will check if the current timestamp is at or over `RenewAt` and will call the `Renew` method by providing the previously returned `Private` information.
+
+> [!NOTE]
+> 
+> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
```

Comment:
When you say "Done using", is that
- when the apply phase is finished
- when the value is no longer referenced by a resource/output?
- when the node in the graph has been read completely

Can you elaborate on what `done` is here please?

Reply:
Yeah, I am having a rough time wording that correctly. "done using" in this context means that the apply phase finished and now it executes the close nodes. Right before executing `graphNodeCloseProvider` nodes, it should call `Close` on the ephemeral resources.
If you can help with a good wording around this, would greatly appreciate it.